### DISCUSSION TITLE
Question on ordering of fields

### DISCUSSION THREAD

**Discussion by CaptainDriftwood** [wyfo/apischema#226] [Q&A] [answered]
Just wondering what it is I am doing incorrectly when wanting to order fields:

Per the documentation (https://wyfo.github.io/apischema/de_serialization/#class-level-ordering), this works as expected.

However, when I try to do the following:
```python
@order(["hairColor", "name", "age"])
@dataclass
class Person:
    """Model to represent a person object"""

    name: str
    age: int
    hair_color: str = field(metadata=alias("hairColor"))

person = Person("John", 21, "Brown")
print(json.dumps(serialize(person)))
>>> '{"hairColor": "Brown"}'
```

What I would expect the output to be is as follows:
```python
>>> '{"hairColor": "Brown", "name": "John", "age": 33}'
```

Am I doing something wrong? Also, when decorating the class with the `order` decorator, type hints get lost when instantiating the class object. Is there any way to fix that?

**Comment by wyfo** [comment, marked as answer]
There are several issue here: a misuse due to an incomplete documentation, a bug, and maybe a feature request.

On the one hand, about the documentation; *apischema* has several features expecting fields as argument, for example [fields validator](https://wyfo.github.io/apischema/validation/#using-other-fields) or ordering. This not really documented, but fields can be passed either as a dataclass field (convenient inside class body), `apischema.objects.ObjectField` or a string, which is expected to be **the field name**. Field alias cannot be used here, so your example will not work. You have to replace `"hairColor"` by `"hair_color"`.

On the other hand, there is a bug in the current implementation (i will fix it today and release it in v0.16.1 at the same time). Ordering `after` and `'before` are not resolved recursively, and applying `order` at class level is just a shortcut to chain `after`. In your example, it would give:
```python
@dataclass
class Person:
    name: str = field(metadata=order(after="hair_color"))
    age: int = field(metadata=order(after="name"))
    hair_color: str = field(metadata=alias("hairColor"))
```
That's why, even if you fix your decorator to be `@order(["hair_color", "name", "age"])`, `age` field will still not be serialized.  The correction is quite trivial, and as I've written above, it will be patched and available today.

I will also update the documentation to precise that field aliases cannot be used, for the moment. I understand that it could be convenient to use aliases in addition to field names, and I will think about it. I should not be so hard to do, even if there is some edge cases to handle, but I have other priority things to implement at this moment.

By the way, regarding your example, *apischema* provides [dynamic/global aliasing](https://wyfo.github.io/apischema/json_schema/#dynamic-aliasing-and-default-aliaser), which is a lot more convenient than aliasing fields one by one, especially for *camelCase* handling.

P.S. Don't hesitate to create issue instead of question when the things looks a lot like a bug (having fields not serialized). I've created #227

**Reply by CaptainDriftwood** [reply]
Awesome, thank you very much for the explanation. I was not sure that my initial finding for the ordering was a bug or just a lack of understanding the implementation. Also, in addition to  the global alias you mentioned, is it possible to do the following:

- define aliasing behavior at the class level instead of global/per field level (via a decorator)?
- define aliasing for a specific serializer (i.e., using one serializer for one use case, and a different one for another)?

**Reply by wyfo** [reply]
> define aliasing behavior at the class level instead of global/per field level (via a decorator)?

Yes, take a look at the [documentation](https://wyfo.github.io/apischema/json_schema/#alias-all-fields)

> define aliasing for a specific serializer (i.e., using one serializer for one use case, and a different one for another)?

I'm not sure to understand what do you mean. If you want for example to serialize with *camelCase* for on one side, and with *snake_case* on the other side, you can use `aliaser` parameter of `apischema.serialize`/`apischema.serialization_method`.
Otherwise, if you mean having two different aliasing for the same class, depending on a given context, you could use [object serialization](https://wyfo.github.io/apischema/conversions/#object-serialization-select-only-a-subset-of-fields) in order to override the `alias` metadata of the fields you want.
(There are also equivalent mechanisms for deserialization)

Does it answer your question?

**Reply by CaptainDriftwood** [reply]
Yes, that does answer my question. Also, with the new `order` decorator, is there a way to have the order of the fields be alphabetical (in either ascending or descending order)? I'm not seeing a way to specify this behavior at the class decorator level in the documentation.

**Reply by wyfo** [reply]
>  is there a way to have the order of the fields be alphabetical

There is not, at least at simple one, and that could indeed be a good feature idea.
If you don't have any [serialized methods](https://wyfo.github.io/apischema/de_serialization/#serialized-methodsproperties), here is a way you can accomplish that with one (long) line of code:
```python
import json
from dataclasses import dataclass

from apischema import serialize, serializer, type_name
from apischema.json_schema import serialization_schema
from apischema.objects import object_fields, object_serialization


@dataclass
class MyClass:
    foo: int
    bar: int


# Create a dynamic object serialization reusing all fields, but sorted;
myclass_sorted = object_serialization(
    MyClass,
    lambda: sorted(object_fields(MyClass).values(), key=lambda f: f.alias),
    type_name(None),  # in order to not have intermediate class added in JSON schema.
)
# Register a default serializer
serializer(myclass_sorted)

assert json.dumps(serialize(MyClass(0, 1))) == '{"bar": 1, "foo": 0}'
assert serialization_schema(MyClass, all_refs=True) == {
    "$ref": "#/$defs/MyClass",
    "$defs": {
        "MyClass": {
            "type": "object",
            "properties": {"bar": {"type": "integer"}, "foo": {"type": "integer"}},
            "required": ["bar", "foo"],
            "additionalProperties": False,
        }
    },
    "$schema": "http://json-schema.org/draft/2020-12/schema#",
}
```

**Reply by wyfo** [reply]
I've created #229 to continue the discussion about alphabetical ordering. Feel free to take a look at it.

**Reply by CaptainDriftwood** [reply]
Awesome, this is good stuff. Thank you for the assistance and the very prompt responsiveness. Out of my own curiosity, how would one implement this with serialized methods/properties (or is this not currently possible)?
